### PR TITLE
Ea youtube render

### LINKED ISSUE #15: Consume Youtube API
---
name: Feature request
about: Displaying related youtube videos on article show page
---

**Is your feature request related to a problem? Please describe.** 

No.

**Describe the solution you'd like**

As a user,
On the article show page, 
I should see three youtube videos related to the tags of the article, 
in an aesthetically pleasing format.

**Describe alternatives you've considered**

Could possibly use keywords for related videos as well. 
 
**Additional context**

User google youtube v3 API

### LINKED ISSUE #17: Create video objects for front-end manipulation
---
name: Feature request
about: Displaying related youtube videos on article show page
---

**Is your feature request related to a problem? Please describe.** 

No.

**Describe the solution you'd like**

As a user,
On the article show page, 
I should see three youtube videos related to the tags of the article, 
in an aesthetically pleasing format.

**Describe alternatives you've considered**

Could possibly use keywords for related videos as well. 

** Additional context**

Elom is going to fetch the consumed youtube API from within the article show page. He will use the route get "/:username/videos/youtube" => "youtube#index" to direct the app to the youtube controller, where the api will then be consumed. We will also utilize a partial erb form to render the videos in a aesthetically pleasing format.

### PR DESCRIPTION
**Pull request by eamouzou**
## What type of PR is this? (check all applicable)

- [ X] Refactor
- [X ] Feature

- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Added youtube video functionality to article show page

## Added tests?

- [X] yes
Test needs to be refactored; seems to be an issue with Webmock and http connections.

- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [X] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: spec/requests/youtube/youtube_request_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      article_1 = create(:article, user_id: user.id)
+      article_2 = create(:article, user_id: user.id)
+      # article_comment = create(:comment, commentable_id: article_1.id, commentable_type: "Article", user_id: user.id)
+      # podcast = create(:podcast)
+      # podcast_episode = create(:podcast_episode, podcast_id: podcast.id)
+      # podcast_comment = create(:comment, commentable_id: podcast_episode.id, commentable_type: "PodcastEpisode", user_id: user.id)
+      last_article = create(:article, user_id: user.id)
+      tags = last_article.cached_tag_list
+
+      get "/#{user.username}/videos/youtube?tags=#{tags}"
```
Comment (eamouzou) [inline spec/requests/youtube/youtube_request_spec.rb:20]: @Will-Meighan ; this test needs to be refactored. When I enable real http connections for webmock, it passes and the functionality works in localhost. However, with the default setting for webmock being no real htttp connections allowed, it fails.

Comment (Will-D-M) [reply spec/requests/youtube/youtube_request_spec.rb:20]: I can take a look at that.

File: config/routes.rb [hunk 2]
```diff
@@ -412,6 +412,9 @@
   get "/:username/:slug" => "stories#show"
   get "/:username" => "stories#index"
 
+  # ea - the fetch call in the youtube partial should come here
+  get "/:username/videos/youtube" => "youtube#index"
```
Comment (Will-D-M) [inline config/routes.rb:416]: I am still a little confused by this route. The route that we agreed would render the videos is get "/:username/:slug" => "stories#show". I trust your work but will have to have you explain this one to me.

File: app/views/articles/_youtube.html.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+function displayVideos(videos) {
+    let videoContainer = document.querySelector('.youtube')
+    if (videos.length === 0) {
+      return videoContainer.innerHTML = `<h2>There were no videos found matching your topic.</h2>`
+    } else {return videos.map(video => {
+        let id = video.id;
+        videoContainer.innerHTML += `<article class="video-article">
+            <%= render "liquids/youtube", id: "${id}", width: 350, height: 350 %>
+          </article>`})}}
+  fetch(`/<%= @user.username %>/videos/youtube?tags=<%= @article.cached_tag_list %>`)
```
Comment (iandouglas) [inline app/views/articles/_youtube.html.erb:12]: basing this only tags might be pretty limiting and would require some refactoring if you were to change this later

File: spec/requests/youtube/youtube_request_spec.rb (hunk 1 above)
Comment (iandouglas) [reply spec/requests/youtube/youtube_request_spec.rb:20]: maybe look into VCR to cache a real response from YouTube and use that going forward?

File: config/routes.rb (hunk 2 above)
Comment (eamouzou) [reply config/routes.rb:416]: No problem. This route is called from the partial we are rendering for the videos in thee articles/show view. It connects to the youtube controller and serializes the data for the youtube videos (from a facade which initializes a service).

File: app/services/youtube_service.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  def conn
+    Faraday.new("https://www.googleapis.com/youtube/v3/search") do |f|
+      f.adapter Faraday.default_adapter
+      f.params[:key] = ENV["YOUTUBE_API_KEY"]
+      f.params[:q] = @tags
+      f.params[:part] = "snippet"
+      f.params[:type] = "video"
+      f.params[:videoEmbeddable] = "true"
+      f.params[:maxResults] = "3"
```
Comment (Will-D-M) [inline app/services/youtube_service.rb:27]: Love the organization here. Should make for easily manipulated object for front end.

**Pull request merged**